guides/docs/testing/testing_controllers.md

```diff
@@ -133,169 +222,489 @@ defmodule HelloWeb.Router do
-    resources "/users", HelloWeb.UserController
   end
 
-  ...
+  scope "/", HelloWeb do
+    pipe_through :browser # Use the default browser stack
+
+    get "/", PageController, :index
+  end
+
+  # Other scopes may use custom stacks.
+  scope "/api", HelloWeb do
+    pipe_through :api
+    resources "/users", UserController, except: [:new, :edit]
+  end
+end
+```
+
+Before running the test again, check out our new paths by running `mix
+phx.routes`.  You should see six new "/api" routes in addition to the default
+page controller route:
+
+```console
+$ mix phx.routes
+Compiling 6 files (.ex)
+page_path  GET     /               HelloWeb.PageController :index
+user_path  GET     /api/users      HelloWeb.UserController :index
+user_path  GET     /api/users/:id  HelloWeb.UserController :show
+user_path  POST    /api/users      HelloWeb.UserController :create
+user_path  PATCH   /api/users/:id  HelloWeb.UserController :update
+           PUT     /api/users/:id  HelloWeb.UserController :update
+user_path  DELETE  /api/users/:id  HelloWeb.UserController :delete
 ```
 
-We should get a new error now. Running the test informs us we don't have a `UserController`. Let's add it, along with the `index/2` action we're testing. Our test description has us returning all users:
+
+
+We should get a new error now. Running the test informs us we don't have a `HelloWeb.UserController`. Let's add it, along with the `index/2` action we're testing. Our test description has us returning all users:
 
 ```elixir
+# lib/hello_web/controllers/user_controller.ex
+
 defmodule HelloWeb.UserController do
   use HelloWeb, :controller
-
-  alias Hello.{Accounts.User, Repo}
+  alias Hello.Accounts
 
   def index(conn, _params) do
-    users = Repo.all(User)
-    render conn, "index.json", users: users
+    users = Accounts.list_users()
+    render(conn, "index.json", data: users)
   end
 
 end
 ```
 
-When we run the test again, our failing test tells us we have no view. Let's add it. Our test specifies a JSON format with a top key of `"data"`, containing an array of users with attributes `"name"` and `"email"`.
+When we run the test again, our failing test tells us module `HelloWeb.UserView` is not available. Let's add it. Our test specifies a JSON format with a top key of `"data"`, containing an array of users with attributes `"name"` and `"email"`.
 
 ```elixir
+# lib/hello_web/views/user_view.ex
+
 defmodule HelloWeb.UserView do
   use HelloWeb, :view
 
-  def render("index.json", %{users: users}) do
-    %{data: render_many(users, HelloWeb.UserView, "user.json")}
+  def render("index.json", %{data: users}) do
+    %{data:
+      render_many( users, HelloWeb.UserView, "user.json", as: :data)
+      }
   end
 
-  def render("user.json", %{user: user}) do
-    %{name: user.name, email: user.email}
+  def render("user.json", %{data: user}) do
+    %{
+      name: user.name,
+      email: user.email
+      # skipping password, inserted_at, and updated_at
```

Comment:
Okay.  I think I originally wrote that because I was also working the other way...editing the autogenerated stuff in which password is added but not needed.  And then (as a newcomer to Phoenix...really, I just started using it about two weeks ago) I'm not at all sure why the timestamps are stashed in the database, what purpose they would serve if they're never echoed back to the UI, etc.  

But sure, I'll delete the comment.

Reply:
fixed